### PR TITLE
Raise error when `modules_to_save` is specified and multiple adapters are being unloaded

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Raise error when `modules_to_save` is specified and multiple adapters are being unloaded

### CODE REVIEW

File: src/peft/tuners/lora/model.py [hunk 1]
```diff
@@ -377,6 +377,15 @@ def _unload_and_optionally_merge(
             if getattr(self.model, "quantization_method", None) == "gptq":
                 raise ValueError("Cannot merge LORA layers when the model is gptq quantized")
 
+        adapters_to_consider = adapter_names or self.active_adapters
+        is_modules_to_save_available = any(
+            self.peft_config[adapter].modules_to_save is not None for adapter in adapters_to_consider
+        )
+        if is_modules_to_save_available and len(adapters_to_consider) > 1:
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/model.py:384]: If `adapter_names` is a str, this will trigger an error even though it would be fine, right?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for stepping up and addressing this issue so quickly.

The same change is needed for `ia3/model.py` and `lycoris_utils.py`. Maybe it can be factored out into a utility function?

For general context: This is an oversight from #1132, which was discussed internally. Since the PR is very recent, this change should not affect any users.

File: src/peft/tuners/lora/model.py (hunk 1 above)
Comment (pacman100) [reply src/peft/tuners/lora/model.py:384]: `adapter_names` is a `List[str]`, right?

File: src/peft/tuners/tuners_utils.py [hunk 2]
```diff
@@ -299,6 +299,16 @@ def unmerge_adapter(self):
             if isinstance(module, BaseTunerLayer):
                 module.unmerge()
 
+    def _unloading_checks(self, adapter_names: List[str]):
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:302]:
```suggestion
    def _unloading_checks(self, adapter_names: Optional[List[str]]):
```

File: src/peft/tuners/tuners_utils.py [hunk 3]
```diff
@@ -299,6 +299,16 @@ def unmerge_adapter(self):
...
+        adapters_to_consider = adapter_names or self.active_adapters
+        is_modules_to_save_available = any(
+            self.peft_config[adapter].modules_to_save is not None for adapter in adapters_to_consider
+        )
+        if is_modules_to_save_available and len(adapters_to_consider) > 1:
+            raise ValueError(
+                "models having `modules_to_save` specified in LoraConfig support only single adapter for unloading."
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:309]: I tried to come up with an error message that is a bit clearer, WDYT?

```suggestion
                "Cannot unload multiple adapters that specify `modules_to_save`."
```

File: src/peft/tuners/tuners_utils.py [hunk 4]
```diff
@@ -299,6 +299,16 @@ def unmerge_adapter(self):
             if isinstance(module, BaseTunerLayer):
                 module.unmerge()
 
+    def _unloading_checks(self, adapter_names: List[str]):
+        adapters_to_consider = adapter_names or self.active_adapters
+        is_modules_to_save_available = any(
+            self.peft_config[adapter].modules_to_save is not None for adapter in adapters_to_consider
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:305]:
```suggestion
            self.peft_config[adapter].modules_to_save for adapter in adapters_to_consider
```

How about this change? This way, if `modules_to_save = []`, it is not counted here.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for making the adjustments. I only have some small comments, the rest looks good.

**Review by BenjaminBossan — APPROVED** [review]
Great, thanks! From my point of view, can be merged as soon as CI has passed.

**Pull request merged**